sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/EventTimeEquiJoin.java

```diff
@@ -0,0 +1,272 @@
+                    secondValueCoder,
+                    getFirstCollectionValidFor(),
+                    getSecondCollectionValidFor(),
+                    getAllowedLateness())))
+        .setCoder(KvCoder.of(keyCoder, PairCoder.<V1, V2>of(firstValueCoder, secondValueCoder)));
+  }
+
+  private static class EventTimeEquiJoinDoFn<K, V1, V2>
+      extends DoFn<KV<K, RawUnionValue>, KV<K, Pair<V1, V2>>> {
+    private static final int FIRST_TAG = 0;
+    private static final int SECOND_TAG = 1;
+
+    // Bucket cleanup timers into TIMER_BUCKET length buckets.
+    private static final long TIMER_BUCKET = Duration.standardMinutes(1).getMillis();
+
+    // How long elements in the first and second collection are valid (can be matched) for.
+    private final Duration firstCollectionValidFor;
+    private final Duration secondCollectionValidFor;
+
+    // How long past the watermark that late elements can show up.
+    private final Duration allowedLateness;
+
+    @StateId("v1Items")
+    private final StateSpec<OrderedListState<V1>> firstCollectionItems;
+
+    @StateId("v2Items")
+    private final StateSpec<OrderedListState<V2>> secondCollectionItems;
+
+    @TimerFamily("cleanupTimers")
+    private final TimerSpec cleanupTimers = TimerSpecs.timerMap(TimeDomain.EVENT_TIME);
+
+    public EventTimeEquiJoinDoFn(
+        Coder<V1> firstValueCoder,
+        Coder<V2> secondValueCoder,
+        Duration firstValidFor,
+        Duration secondValidFor,
+        Duration allowedLateness) {
+      this.firstCollectionValidFor = firstValidFor;
+      this.secondCollectionValidFor = secondValidFor;
+      this.allowedLateness = allowedLateness;
+      firstCollectionItems = StateSpecs.orderedList(firstValueCoder);
+      secondCollectionItems = StateSpecs.orderedList(secondValueCoder);
+    }
+
+    @FunctionalInterface
+    private interface Output<T1, T2> {
+      void apply(T1 one, T2 two);
+    }
+
+    private <T, O> void processHelper(
+        Output<T, O> output,
+        KV<K, RawUnionValue> element,
+        Instant ts,
+        OrderedListState<T> thisCollection,
+        OrderedListState<O> otherCollection,
+        TimerMap cleanupTimers,
+        Duration thisCollectionValidFor,
+        Duration otherCollectionValidFor) {
+      thisCollection.add(TimestampedValue.of((T) element.getValue().getValue(), ts));
+      Instant beginning = ts.minus(otherCollectionValidFor);
+      Instant end = ts.plus(thisCollectionValidFor).plus(1L);
+      for (TimestampedValue<O> value : otherCollection.readRange(beginning, end)) {
+        output.apply((T) element.getValue().getValue(), value.getValue());
+      }
+      Instant cleanupTime = ts.plus(allowedLateness).plus(thisCollectionValidFor);
+      Instant nextBucketStart =
+          Instant.ofEpochMilli(
+              cleanupTime.getMillis() / TIMER_BUCKET * TIMER_BUCKET + TIMER_BUCKET);
+      cleanupTimers.get(Long.toString(nextBucketStart.getMillis())).set(nextBucketStart);
```

Comment:
ideally we should call withOutputTimestamp to hold the watermark back to the earliest buffered element in that timer's range. Doing this probably requires keeping a histogram in a ValueState tracking the min timestamp per minute bucket.